header-rewriter/SourceRewriter.cpp

```diff
@@ -0,0 +1,920 @@
+
+    // Depending on which matcher hit, determine what the fn ptr type is and how
+    // to make a new, wrapped declaration of the same kind from it
+    if (auto *parm_var_decl =
+            result.Nodes.getNodeAs<clang::ParmVarDecl>("fnPtrParam")) {
+      old_decl = llvm::cast<clang::Decl>(parm_var_decl);
+      old_type = parm_var_decl->getOriginalType();
+      generate_decl = append_name_if_nonempty;
+    } else if (auto *field_decl =
+                   result.Nodes.getNodeAs<clang::FieldDecl>("fnPtrField")) {
+      old_decl = llvm::cast<clang::Decl>(field_decl);
+      old_type = field_decl->getType();
+      generate_decl = append_name_if_nonempty;
+    } else if (auto *typedef_decl =
+                   result.Nodes.getNodeAs<clang::TypedefDecl>("fnPtrTypedef")) {
+      old_decl = llvm::cast<clang::Decl>(typedef_decl);
+      old_type = typedef_decl->getUnderlyingType();
+      generate_decl = [](const auto &new_type, const auto &name) {
+        return "typedef " + new_type + ' ' + name;
+      };
+    } else if (auto *type_alias_decl =
+                   result.Nodes.getNodeAs<clang::TypeAliasDecl>(
+                       "fnPtrTypedef")) {
+      old_decl = llvm::cast<clang::Decl>(type_alias_decl);
+      old_type = typedef_decl->getUnderlyingType();
+      generate_decl = [](const auto &new_type, const auto &name) {
+        return "using "s + name + " = " + new_type;
+      };
+    } else if (auto *var_decl =
+                   result.Nodes.getNodeAs<clang::VarDecl>("fnPtrVar")) {
+      old_decl = llvm::cast<clang::Decl>(var_decl);
+      old_type = var_decl->getType();
+      // When function pointer variables are initialized
+      // (e.g. int(*ptr)(int) = INITIAL_VALUE;) the initial value will be part
+      // of the SourceRange. Since the FnPtrExpr pass may need to rewrite
+      // INITIAL_VALUE, we reduce the SourceRange to only include the LHS.
+      if (var_decl->hasInit()) {
+        // VarDecl doesn't expose a way to get the SourceRange without the
+        // initializer, so this is taken almost verbatim from
+        // DeclaratorDecl::getSourceRange
+        auto start = var_decl->getOuterLocStart();
+        // The type extends past the variable name (i.e. `ptr` above) so we need
+        // to manually set the SourceRange end
+        auto *ty_info = var_decl->getTypeSourceInfo();
+        assert(ty_info != nullptr);
+        auto end = ty_info->getTypeLoc().getSourceRange().getEnd();
+        range = clang::SourceRange(start, end);
+      }
+      generate_decl = append_name_if_nonempty;
+    }
+    // Get the node's SourceRange if we are not just rewriting a subset of it
+    if (!range.has_value()) {
+      range = old_decl->getSourceRange();
+    }
+
+    assert(result.SourceManager != nullptr);
+    assert(result.Context != nullptr);
+    assert(old_decl != nullptr);
+
+    const auto &sm = *result.SourceManager;
+    auto &ctxt = *result.Context;
+
+    if (get_file_pkey(sm) == 0) {
```

Comment:
Is zero not a valid pkey here?

Reply:
Rewriter passes that modify source files (currently all except FnDecl) should change translation units with pkey 0. I'll add a comment to these